game-core/src/main/java/games/strategy/engine/auto/update/EngineVersionCheck.java

```diff
@@ -55,18 +55,16 @@ static boolean isEngineUpdateCheckRequired(
       final GameSetting<String> updateCheckDateSetting,
       final Runnable flushSetting) {
     // check at most once per 2 days (but still allow a 'first run message' for a new version of TripleA)
-    final boolean firstRun = firstRunSetting.value();
-    final String encodedUpdateCheckDate = updateCheckDateSetting.value();
-    if (!firstRun && !encodedUpdateCheckDate.trim().isEmpty()) {
-      final LocalDate updateCheckDate = parseUpdateCheckDate(encodedUpdateCheckDate);
-      if (updateCheckDate.isAfter(now.minusDays(2))) {
-        return false;
-      }
+    final boolean updateCheckRequired = firstRunSetting.value()
+        || updateCheckDateSetting.getValue()
+            .map(encodedUpdateCheckDate -> !parseUpdateCheckDate(encodedUpdateCheckDate).isAfter(now.minusDays(2)))
```

Comment:
The negation smell here is due to `LocalDate` not offering an `isBeforeOrEqual()` method.

Reply:
Yup, I did consider that, but the lambda expression changed to a very busy lambda block:

```java
|| updateCheckDateSetting.getValue()
    .map(encodedUpdateCheckDate -> {
      final LocalDate updateCheckDate = parseUpdateCheckDate(encodedUpdateCheckDate);
      final LocalDate targetDate = now.minusDays(2);
      return updateCheckDate.isBefore(targetDate) || updateCheckDate.isEqual(targetDate);
    })
    .orElse(true);
```

Which I then extracted to a method:

```java
|| updateCheckDateSetting.getValue()
    .map(encodedUpdateCheckDate -> isBeforeOrEqual(parseUpdateCheckDate(encodedUpdateCheckDate), now.minusDays(2)))
    .orElse(true);
```

Where `isBeforeOrEqual()` is

```java
private static boolean isBeforeOrEqual(final LocalDate date, final LocalDate otherDate) {
  return date.isBefore(otherDate) || date.isEqual(otherDate);
}
```

At the end of that exercise, I figured the negation smell was better than the verbosity.  But I'm happy to change to one of those two alternatives if you find them preferable.

Reply:
I'm OK with a negated statement ^^
And creating a helper method just for an or expression isn't worth it IMO.